compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs

```diff
@@ -227,41 +237,57 @@ impl IntRange {
         }
 
         // The boundaries of ranges in `column_ranges` intersected with `self`.
-        let mut boundaries: Vec<IntBoundary> = column_ranges
+        // We do parenthesis matching for input ranges. A boundary counts as +1 if it starts
+        // a range and -1 if it ends it. When the count is > 0 between two boundaries, we
+        // are within an input range.
+        let mut boundaries: Vec<(IntBoundary, isize)> = column_ranges
             .filter_map(|r| self.intersection(&r))
             .map(unpack_intrange)
-            .flat_map(|[lo, hi]| [lo, hi])
+            .flat_map(|[lo, hi]| [(lo, 1), (hi, -1)])
             .collect();
         boundaries.sort_unstable();
 
+        // Counter for parenthesis matching.
+        let mut paren_counter = 0isize;
+        let boundaries_with_paren_counts = boundaries
+            .into_iter()
+            // Accumulate parenthesis counts.
+            .map(move |(bdy, delta)| {
+                paren_counter += delta;
+                (bdy, paren_counter)
+            });
```

Comment:
Oh I didn't know about this optimization! How would this simplify the logic below?

Reply:
Ah I saw your other comment. I can't use `array_windows` because I need to take ownership of the vec into the iterator. And `dedup` doesn't do what I want: I need to keep the last of each run of the same boundary, not the first. I did try to clarify a bit, lmk